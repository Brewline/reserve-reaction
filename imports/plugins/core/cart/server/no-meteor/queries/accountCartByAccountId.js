import ReactionError from "@reactioncommerce/reaction-error";

/**
 * @name accountCartByAccountId
 * @method
 * @memberof Cart/NoMeteorQueries
 * @summary Query the Cart collection for a cart with the provided accountId and shopId
 * @param {Object} context -  an object containing the per-request state
 * @param {Object} params - request parameters
 * @param {String} [params.accountId] - An account ID
 * @param {String} [params.shopId] - A shop ID
 * @return {Promise<Object>|undefined} A Cart document, if one is found
 */
export default async function accountCartByAccountId(context, { accountId, shopId } = {}) {
  const { accountId: contextAccountId, collections, userHasPermission } = context;
  const { Cart } = collections;

  if (accountId !== contextAccountId && !userHasPermission(["owner", "admin"], shopId)) {
    throw new ReactionError("access-denied", "Access Denied");
  }

  if (!accountId) {
    throw new ReactionError("invalid-param", "You must provide accountId");
  }
  if (!shopId) {
    throw new ReactionError("invalid-param", "You must provide shopId");
  }

  return Cart.findOne({ accountId, shopId });
}
