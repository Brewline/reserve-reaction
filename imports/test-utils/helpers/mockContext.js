const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
