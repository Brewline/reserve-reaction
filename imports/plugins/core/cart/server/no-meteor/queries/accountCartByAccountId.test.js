import mockContext from "/imports/test-utils/helpers/mockContext";
import accountCartByAccountId from "./accountCartByAccountId";

const shopId = "shopId";

test("for logged in account, expect to return a Promise that resolves to a cart", async () => {
  const cart = { _id: "cart" };
  const { accountId } = mockContext;
  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(cart));

  const result = await accountCartByAccountId(mockContext, { accountId, shopId });
  expect(result).toEqual(cart);
  expect(mockContext.collections.Cart.findOne).toHaveBeenCalledWith({ accountId, shopId });
});

test("for other account, allows if admin", async () => {
  const cart = { _id: "cart" };
  const accountId = "123";
  mockContext.userHasPermission.mockReturnValueOnce(true);
  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(cart));

  const result = await accountCartByAccountId(mockContext, { accountId, shopId });
  expect(result).toEqual(cart);
  expect(mockContext.collections.Cart.findOne).toHaveBeenCalledWith({ accountId, shopId });
});

test("for other account, throws access denied if non-admin", async () => {
  const cart = { _id: "cart" };
  const accountId = "123";
  mockContext.userHasPermission.mockReturnValueOnce(false);
  mockContext.collections.Cart.findOne.mockReturnValueOnce(Promise.resolve(cart));

  expect(accountCartByAccountId(mockContext, { accountId, shopId })).rejects.toMatchSnapshot();
});
