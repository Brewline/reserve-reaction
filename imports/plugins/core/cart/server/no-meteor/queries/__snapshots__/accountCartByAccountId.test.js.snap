// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`for other account, throws access denied if non-admin 1`] = `[ReactionError: Access Denied]`;
